src/main/java/org/gaul/s3proxy/S3ProxyHandler.java

```diff
@@ -2871,4 +2873,38 @@ private static boolean validateIpAddress(String string) {
         }
         return true;
     }
+
+    private static final class StandardDataStream implements WriteListener {
+        private static final Logger logger = LoggerFactory.getLogger(
+                StandardDataStream.class);
+
+        private final AsyncContext async;
+        private final InputStream is;
+        private final ServletOutputStream sos;
+
+        private StandardDataStream(AsyncContext async, InputStream is,
+                ServletOutputStream sos) {
+            this.async = async;
+            this.is = is;
+            this.sos = sos;
+        }
+
+        public void onWritePossible() throws IOException {
+            byte[] buffer = new byte[4096];
+            while (sos.isReady()) {
+                int len = is.read(buffer);
```

Comment:
I would like to see the benchmarks that show that this approach is significantly worse in real-world tests.  You will need to do all the legwork on this.  Please update jclouds-dev with your results.